File: 00_programmers/lv2/가장큰수.cpp
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

bool comp(int& s1, int& s2) {
	int s1_len = to_string(s1).length();
	int s2_len = to_string(s2).length();
	int s1s2 = s1 * pow(10, s2_len) + s2;
	int s2s1 = s2 * pow(10, s1_len) + s1;
	// s1s2가 더 크면 왼쪽(먼저 정렬되어야)에 위치해야 함
	return s1s2 > s2s1;
}

string solution(vector<int> numbers) {
	string answer = "";
	sort(numbers.begin(), numbers.end(), comp);
	// 0이 맨 왼쪽에 오면 안됨
	int k = 0;
	while (k < numbers.size() && numbers[k] == 0)
		++k;
	// 모든 숫자가 0인 경우
	if (k == numbers.size())
		answer = "0";
	for (int i = k; i < numbers.size(); ++i)
		answer += to_string(numbers[i]);

	return answer;
}

File: 00_programmers/lv2/가장큰수.java
import java.util.*;

// 두 아들은 같은 양의 과자를 받아야 한다.

class Solution {
    public String solution(int[] numbers) {
        String answer = "";

        Integer[] nums = new Integer[numbers.length];
        for (int i = 0; i < numbers.length; ++i) {
            nums[i] = numbers[i];
        }

        Comparator<Integer> comp = new Comparator<>() {
            @Override
            public int compare(Integer o1, Integer o2) {
                int s1_len = o1.toString().length();
                int s2_len = o2.toString().length();
                int s1s2 = (int) (o1 * Math.pow(10, s2_len) + o2);
                int s2s1 = (int) (o2 * Math.pow(10, s1_len) + o1);
                return s2s1 - s1s2;
            }
        };

        Arrays.sort(nums, comp);
        
        int k = 0;
        while(k < nums.length && nums[k] == 0)
            ++k;
        
        if(k == nums.length)
            answer = "0";
        
        for(int i = k; i < nums.length; ++i)
            answer += nums[i].toString();

        return answer;
    }
}